File: tool/map-editor-cs/MapEditor/Model/MapHistory.cs
using System.Collections.Generic;

namespace tool.mapeditor.model;

public class MapHistory
{
    private readonly Stack<EditableL1Map> _undo = new();
    private readonly Stack<EditableL1Map> _redo = new();

    public void PushSnapshot(EditableL1Map map)
    {
        _undo.Push(map.DeepCopy());
        _redo.Clear();
    }

    public EditableL1Map? Undo(EditableL1Map current)
    {
        if (_undo.Count == 0)
        {
            return null;
        }

        _redo.Push(current.DeepCopy());
        return _undo.Pop();
    }

    public EditableL1Map? Redo(EditableL1Map current)
    {
        if (_redo.Count == 0)
        {
            return null;
        }

        _undo.Push(current.DeepCopy());
        return _redo.Pop();
    }

    public void Clear()
    {
        _undo.Clear();
        _redo.Clear();
    }
}

File: tool/map-editor/src/main/java/tool/mapeditor/model/MapHistory.java
package tool.mapeditor.model;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Simple undo/redo stack for {@link EditableL1Map} instances.
 */
public class MapHistory {
    private final Deque<byte[][]> undoStack = new ArrayDeque<>();
    private final Deque<byte[][]> redoStack = new ArrayDeque<>();

    public void snapshot(EditableL1Map map) {
        undoStack.push(map.copyTiles());
        redoStack.clear();
    }

    public boolean canUndo() {
        return !undoStack.isEmpty();
    }

    public boolean canRedo() {
        return !redoStack.isEmpty();
    }

    public void undo(EditableL1Map map) {
        if (!canUndo()) {
            return;
        }
        redoStack.push(map.copyTiles());
        byte[][] previous = undoStack.pop();
        map.overwriteTiles(previous);
    }

    public void redo(EditableL1Map map) {
        if (!canRedo()) {
            return;
        }
        undoStack.push(map.copyTiles());
        byte[][] next = redoStack.pop();
        map.overwriteTiles(next);
    }

    public void clear() {
        undoStack.clear();
        redoStack.clear();
    }
}
